### PR TITLE
Add support for anchor links

### LINKED ISSUE #1039: Add Support for Anchor Links
### Description

Apple News supports anchor links (e.g., `#my-cool-heading`) but this plugin does not. It intentionally converts various types of referential links to absolute URLs so that they don't break when sent to Apple News (e.g., a relative link like `/my-example-page` would be rewritten to include the domain, like `https://www.example.org/my-example-page`). It would be good to remove this behavior for anchor links and instead make anchor links work properly. In order to do that, we will need to add an `identifier` property to components based on the `id` of the HTML element.

ref: https://developer.apple.com/documentation/apple_news/apple_news_format/components/using_html_with_apple_news_format

### Use Case

> As an editor, I want to be able to create anchor links within my document that work on Apple News as well as natively on my site.

### Acceptance Criteria

- Remove the logic that adds the permalink to the anchor link, leaving the anchor link intact.
- Add parsing for IDs and add IDs as identifiers to components so that they can be accessed by anchor links. Use the first (top-level) ID found within component HTML as the identifier for the component.

### PR DESCRIPTION
**Pull request by attackant**
Fixes #1039 

This PR does a few things:

- Adds ability to add anchor links to HTML elements in body and text components
- Handles validation and deduplication of IDs
- Fixes issue where non static method was called statically `( new BC_Setup() )->action_init();`
- Fixes array to string conversion error
- Extracts HTML cleanup into multiple methods
- Updates unit tests
- Fixes an edge case where `Apple_News:is_initialized()` never returns true
- Incremental code cleanup and modernization
- Adding some type checking to private methods

### CODE REVIEW

File: includes/apple-exporter/builders/class-components.php [hunk 1]
```diff
@@ -426,10 +457,69 @@ private function clean_up_components( &$component ) {
...
+				if ( preg_match( '/^\d/', $matches[1] ) ) {
+					return '';
+				}
+
+				// Saving the 'id' as the 'identifier'.
+				$identifier = $matches[1];
+
+				// If this identifier already exists skip it (is a duplicate).
+				if ( isset( $identifiers[ $identifier ] ) ) {
+					return '';
```
Comment (mogmarsh) [inline includes/apple-exporter/builders/class-components.php:500]: do you want to increment the count here? Is the count used for anything? If not, maybe just take that out of the comments.

File: includes/apple-exporter/class-parser.php [hunk 2]
```diff
@@ -142,72 +138,110 @@ private function parse_markdown( $html ) {
...
+	}
+
+	/**
+	 * Remove empty <a> tags from the given HTML content.
+	 *
+	 * @param string $html The HTML content to remove empty <a> tags from.
+	 *
+	 * @return string The modified HTML content without empty <a> tags.
+	 */
+	private function remove_empty_tags( string $html ): string {
```
Comment (mogmarsh) [inline includes/apple-exporter/class-parser.php]: should this be `remove_empty_a_tags` to be a little more descriptive?

File: includes/apple-exporter/class-parser.php [hunk 3]
```diff
@@ -142,72 +138,110 @@ private function parse_markdown( $html ) {
...
+	 * Replace the root-relative URLs with the absolute
+	 * URLs using the site URL.
+	 *
+	 * @param string $html The HTML content to handle root-relative URLs.
+	 *
+	 * @return string The modified HTML content with absolute URLs for root-relative ones.
+	 */
+	private function handle_root_relative_urls( string $html ): string {
+		return preg_replace_callback(
+			'/(<a[^>]+href="\/[^\/"][^"]*")[^>]*>/m',
```
Comment (mogmarsh) [inline includes/apple-exporter/class-parser.php]: I feel like `[^\/"]` should be `[^\/]`

File: includes/apple-exporter/components/class-body.php [hunk 4]
```diff
@@ -476,12 +482,12 @@ private function dropcap_determination( $theme, $html ) {
 		/**
 		 * Allows for filtering of the dropcap content before return.
 		 *
-		 * @since 2.4.0
-		 *
 		 * @param bool                  $use_dropcap Whether to apply a dropcap to this paragraph or not.
 		 * @param string                $html The post content to filter.
-		 * @param \Apple_Exporter\Theme $theme The theme whose dropcap options are used.
+		 * @param Theme $theme The theme whose dropcap options are used.
 		 * @param string                $post_id The id of the post whose content we're parsing.
```
Comment (mogmarsh) [inline includes/apple-exporter/components/class-body.php]: do we need to align all the param names now that `\Apple_Exporter\` has been removed?

File: includes/apple-exporter/components/class-component.php [hunk 5]
```diff
@@ -384,7 +402,7 @@ public function anchor() {
 	 * @access public
 	 */
 	public function is_anchor_target() {
-		return ! is_null( $this->uid );
+		return '' !== $this->uid;
```
Comment (mogmarsh) [inline includes/apple-exporter/components/class-component.php]: would `! empty()` be good? That would check for `''` and `null`

File: includes/apple-exporter/components/class-component.php [hunk 6]
```diff
@@ -394,7 +412,7 @@ public function is_anchor_target() {
 	 * @access public
 	 */
 	public function can_be_anchor_target() {
-		return $this->can_be_anchor_target && is_null( $this->uid );
+		return $this->can_be_anchor_target && '' === $this->uid;
```
Comment (mogmarsh) [inline includes/apple-exporter/components/class-component.php]: same

File: includes/apple-exporter/components/class-component.php [hunk 7]
```diff
@@ -404,7 +422,7 @@ public function can_be_anchor_target() {
 	 * @access public
 	 */
 	public function uid() {
-		if ( is_null( $this->uid ) ) {
+		if ( '' === $this->uid ) {
```
Comment (mogmarsh) [inline includes/apple-exporter/components/class-component.php]: same

File: includes/apple-exporter/components/class-component.php [hunk 8]
```diff
@@ -416,16 +434,17 @@ public function uid() {
...
+	 *
 	 * @return string The URL to use for this asset in the JSON.
 	 */
 	protected function maybe_bundle_source( $source, $filename = null ) {
 		if ( 'yes' === $this->get_setting( 'use_remote_images' ) ) {
 			return $source;
 		} else {
-			if ( null === $filename ) {
-				$filename = \Apple_News::get_filename( $source );
+			if ( '' === $filename ) {
```
Comment (mogmarsh) [inline includes/apple-exporter/components/class-component.php:446]: maybe use `empty()` here too?

File: includes/class-apple-news.php [hunk 9]
```diff
@@ -1132,7 +1142,7 @@ protected function load_example_themes() {
 	}
 
 	/**
-	 * Verifies that the list of keys provided all exist in the settings array.
+	 * Verifies that the list of keys provided all exists in the settings array.
```
Comment (mogmarsh) [inline includes/class-apple-news.php]: I would go back to `exist` or maybe change it to `Verifies that the keys provided all exist in the settings array.` so there's less confusion.

**Review by mogmarsh — APPROVED** [review]
This looks really good. I made a bunch of comments, but for the most part, they're all suggestions. Feel free to disagree with me and ignore them. 👍

File: includes/apple-exporter/builders/class-components.php (hunk 1 above)
Comment (attackant) [reply includes/apple-exporter/builders/class-components.php:500]: Good call, originally I was going to modify the identifiers to be name-1, name-2, but realized that didn't make sense since people wouldn't be able to target them in WordPress. I'll remove.

File: includes/apple-exporter/class-parser.php (hunk 3 above)
Comment (attackant) [reply includes/apple-exporter/class-parser.php]: I updated the regex slightly to work with both `"/someurl"` and `'/someurl'`

File: includes/apple-exporter/components/class-component.php (hunk 5 above)
Comment (attackant) [reply includes/apple-exporter/components/class-component.php]: Good call, I was using a string before because I was trying to figure out how the variable was changing through the app, but empty is likely a better solution. Will change.

File: includes/apple-exporter/components/class-component.php (hunk 6 above)
Comment (attackant) [reply includes/apple-exporter/components/class-component.php]: Done

**Comment by attackant** [comment]
Thanks @mogmarsh I went ahead with all of your suggestions. Will merge this after I make another round of update for the Heading component. 🥳
